CARDS-1892 / CARDS-1894 / CARDS-1896 / CARDS-1897 / CARDS-1898 / CARDS-1899: Send notifications to Slack whenever important security updates are available for a CARDS deployment

IntegralProgrammer opened the pull request:
**This PR contains the `CARDS-1892`, `CARDS-1896`, `CARDS-1894`, `CARDS-1897`, `CARDS-1898` and `CARDS-1899` implemented JIRA tasks, therefore testing this PR will test all of these tasks. It implements a large part of CARDS-1701 excluding only CARDS-1893 (already merged into `dev`) and CARDS-1895**

To test this PR, follow the instructions in the `Utilities/SecurityScanning/README.md` file and ensure that you are able to receive the full security scan report on GitHub and a summary of the report on Slack. During testing, for the `SLACK_INCOMING_WEBHOOK_URL`, if you do not wish to setup a basic Slack App with an incoming webhook URL, please send me a message on Slack and I can add you to a _testing_ Slack alerts channel and provide you with its incoming webhook URL.

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Add the Utilities/Administration/github_publish_os_package_list.py script

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the build_docker_image.sh script which builds a very minimal Docker image (FROM scratch) and simply copies in the files necessary for Trivy to perform a vulnerablity scan of the APT/DPKG packages installed on a host

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the trivy_to_slack.py script which generates a Slack message block that lists the vulnerable packages detected by Trivy based on its JSON output

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Code cleanup in trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add support for the --package_emoji flag to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the github_download_deployment_packages_config.py script which uses the GitHub App API to download the contents of a deployment_hostname directory in a GitHub repository

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add support for specifying the GitHub API private key, GitHub API App ID, GitHub API installation ID, and GitHub repository (OWNER / REPO-NAME) via environment variables for the github_download_deployment_packages_config.py script

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the scan_remote_package_list.sh script

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Fix permission bits - make build_docker_image.sh scan_remote_package_list.sh executable by the owner

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Improve error handling in the build_docker_image.sh and scan_remote_package_list.sh scripts

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the generate_slack_vulnerability_report_for_host.py script which adds the hostname header and *VM OS* sub-title for a Slack-formatted listed of vulnerabilities provided in the JSON file specified by the --vm_os_json_file argument

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add the post_message_blocks_to_slack.py script which posts a set of Slack message blocks (in JSON format) to the Slack incoming webhook specified by the SLACK_INCOMING_WEBHOOK_URL environment variable

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Address code review comment by de-duplicating calls to updateFile through the use of the OS_FILE_SET map defined in the os_file_management.py file

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add the get_vulnerabilities_for_mvn_pkg.sh script which receives the Maven groupId, artifactId, and version and uses Trivy to determine if known vulnerabilities with fixes exist for the specified Maven package

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Change the output of get_vulnerabilities_for_mvn_pkg.sh to JSON

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Move get_vulnerabilities_for_mvn_pkg.sh and update_trivy_cache.sh into the trivy-utils directory

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

In get_vulnerabilities_for_mvn_pkg.sh, delete the generated pom.xml file once we are done with it

IntegralProgrammer committed:
CARDS-1894: Publish a list of all Maven artifacts that are included in each CARDS Docker image that is in deployment

Add utilities to list all Maven artifacts that are included in a Docker image

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add license headers to update_trivy_cache.sh and get_vulnerabilities_for_mvn_pkg.sh

IntegralProgrammer committed:
CARDS-1894: Publish a list of all Maven artifacts that are included in each CARDS Docker image that is in deployment

Search relative to /root/.m2/repository instead of /root/.m2

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Port the functionality of get_vulnerabilities_for_mvn_pkg.sh to Python

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add the scan_maven_package_list.py script which reads a JSON file of Maven packages, uses get_vulnerabilities_for_mvn_pkg.getVulnerabilities to determine if any packages within that list are vulnerable, and outputs a JSON-encoded Slack message block to be used in a vulnerability report

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add support for --maven_json_file to generate_slack_vulnerability_report_for_host.py

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add support for the --truncate_results parameter to scan_maven_package_list.py as the Slack API has a message length limit

Always list vulnerabilities in order of severity: CRITICAL, HIGH, MEDIUM, LOW

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Begin work on the slack_vulnerability_report_job.sh script

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add VM OS package scanning to the slack_vulnerability_report_job.sh script

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add Docker container OS package scanning to the slack_vulnerability_report_job.sh script

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add support for the --docker_os_json_file parameter to the generate_slack_vulnerability_report_for_host.py script

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Add the github_download_packages_config.py script which downloads the yarn.lock file associated with a Docker image that is deployed on a host

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Add --npm_json_file support to generate_slack_vulnerability_report_for_host.py

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Add the scan_remote_npm_list.sh script which downloads a yarn.lock file from GitHub, uses Trivy to analyze it for vulnerabilities, and formats the results to be displayed in a Slack message

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add NPM package scanning to the slack_vulnerability_report_job.sh script

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Use the :octocat: emoji when listing vulnerable NPM/Yarn packages

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Make the update_trivy_cache.sh script executable

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Cache the Trivy vulnerability database

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Overwrite SLACK_MESSAGES_FILE instead of appending to it

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Cache the Trivy vulnerability database

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Cache the Trivy vulnerability database when scanning the package listing Docker image

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add the scan_remote_maven_list.sh script which downloads the list of Maven packages to scan from GitHub and performs a scan using Trivy

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Change the path on GitHub from DEPLOYMENT_HOSTNAME/vm/... to hosts/DEPLOYMENT_HOSTNAME/vm/...

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Use the scan_remote_maven_list.sh for scanning for Maven package vulnerabilities

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Download package configuration files from hosts/DEPLOYMENT_HOSTNAME/vm instead of from DEPLOYMENT_HOSTNAME/vm

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Download package configuration files from hosts/DEPLOYMENT_HOSTNAME/vm instead of from DEPLOYMENT_HOSTNAME/vm

Fix hosts --> hosts/

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add the scan_remote_maven_list.sh script which downloads the list of Maven packages to scan from GitHub and performs a scan using Trivy

Look under hosts/DEPLOYMENT_HOSTNAME/docker/cards instead of under DEPLOYMENT_HOSTNAME/docker/cards

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Fix github_download_yarn_packages_config.py to reflect DEPLOYMENT_HOSTNAME --> hosts/DEPLOYMENT_HOSTNAME file structure update

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Replace cardsdemo with cardsdemo.uhndata.io for DEPLOYMENT_HOSTNAME argument to scan_remote_package_list.sh

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Removed hardcoded reference to 0.9.5 tag of cards-apk-listing Docker image and replace it with the value specified in the GitHub repository under hosts/cardsdemo.uhndata.io/docker/cards

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Parameterize the slack_vulnerability_report_job.sh script

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add support for the --truncate_results parameter to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add documentation for the --truncate_results parameter to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Add support for the --markdown_report_file parameter to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Pass extra arguments given to scan_remote_package_list.sh to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Pass extra arguments given to scan_remote_maven_list.sh to scan_maven_package_list.py

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Pass extra arguments given to scan_remote_npm_list.sh to trivy_to_slack.py

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Update slack_vulnerability_report.sh to support newly introduced parameters in the dependency scripts

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Add support for the --markdown_report_file parameter to scan_maven_package_list.py

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Update slack_vulnerability_report_job.sh to support --markdown_report_file parameter to scan_remote_maven_list.sh

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Generate the Markdown-formatted security scan report to be posted to GitHub

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add the github_publish_scan_report.py script

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Push the complete Markdown-formatted security scan report to GitHub at the end of slack_vulnerability_report_job.sh

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add a message to the posted Slack block pointing to the location of the full security scan report

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Refactor github_publish_os_package_list.py by moving the GitHub API interaction code into GitHubRepoHandler.py as future scripts to be created under CARDS-1701 will need to use this functionality

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Refactor github_download_deployment_packages_config.py so that all GitHub API interaction code is done through GitHubRepoHandler.py

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Refactor github_download_maven_packages_config.py so that all interaction with the GitHub API is done via GitHubRepoHandler.py

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Refactor github_download_yarn_packages_config.py so that all interaction with the GitHub API is done via GitHubRepoHandler.py

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Refactor github_publish_scan_report.py so that all interaction with the GitHub API is done through GitHubRepoHandler.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Refactor trivy_to_slack.py so that the core logic of Slack message and Markdown docuement generation can be used from external Python scripts in the form of a module import

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Refactor scan_maven_package_list.py to use TrivyToSlackConverter

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Rename github_download_maven_packages_config.py to github_download_docker_related_resource.py and add the --resource_name parameter so that it can be used to download any file under docker/DOCKER_IMAGE_ID and not just docker/DOCKER_IMAGE_ID/maven.json

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Remove github_download_yarn_packages_config.py and do the same operation with github_download_docker_related_resource.py

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Display the installed version of the vulnerable package in the GitHub Markdown report

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Optimize formatting/verbosity of Slack messages

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1896: Write a script that takes a set of APT or APK package status files and determines if there are available fixes for any vulnerabilities within those packages

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1894: Publish a list of all Maven artifacts that are included in each CARDS Docker image that is in deployment

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1897: Write a script that takes a list of Maven artifacts and determines if there are available fixes for any vulnerabilities within those artifacts

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1898: Write a script that takes a yarn.lock file and determines if there are available fixes for any vulnerabilities within the packages listed in said file

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Move to Utilities/SecurityScanning

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Add a README.md file

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Fix install_os_pkg_app.png

sdumitriu commented on Utilities/SecurityScanning/github_download_docker_related_resource.py:
```diff
@@ -0,0 +1,58 @@
+from GitHubRepoHandler import GitHubRepoHandler
+
+argparser = argparse.ArgumentParser()
+argparser.add_argument('--private_key', help='File path the to private key used for JWT signing')
```
```suggestion
argparser.add_argument('--private_key', help='File path to the private key used for JWT signing')
```

sdumitriu committed:
[misc] Typo fix

sdumitriu commented on Utilities/SecurityScanning/github_get_docker_image_digest.py:
```diff
@@ -0,0 +1,114 @@
+else:
+	REPOSITORY = args.repository
+
+def readGitHubTextFile(path, installation_token):
```
Can't this use the `GitHubRepoHandler`?

IntegralProgrammer replied:
Yes, I've made that code simplification in https://github.com/data-team-uhn/cards/pull/1179/commits/ceabcc98e4bbb96bdec624fd2c0fd936df03564d. :heavy_check_mark:

sdumitriu commented on Utilities/SecurityScanning/slack_vulnerability_report_job.sh:
```diff
@@ -0,0 +1,117 @@
+./trivy-utils/update_trivy_cache.sh
+
+# Scan for VM OS vulnerabilities
+# TODO
```
What are all these TODOs for?

IntegralProgrammer replied:
Addressed in https://github.com/data-team-uhn/cards/pull/1179/commits/008385b6eee93bd1576a21a972310412774b39df :heavy_check_mark:

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Simplify code in github_get_docker_image_digest.py by using the GitHubRepoHandler.readGitHubTextFile() method

IntegralProgrammer committed:
CARDS-1899: Write a script that runs all vulnerability check scripts and publishes a summary of the results to a Slack channel

Remove TODO comments in slack_vulnerability_report_job.sh as these scans have now been implemented
